x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/editor_frame.tsx

```diff
@@ -35,28 +36,78 @@ export function EditorFrame(props: EditorFrameProps) {
   const [state, dispatch] = useReducer(reducer, props, getInitialState);
   const { onError } = props;
 
-  // create public datasource api for current state
-  // as soon as datasource is available and memoize it
-  const datasourcePublicAPI = useMemo(
-    () =>
-      state.datasource.activeId && !state.datasource.isLoading
-        ? props.datasourceMap[state.datasource.activeId].getPublicAPI(
-            state.datasource.state,
-            (newState: unknown) => {
+  const allLoaded = Object.values(state.datasourceStates).every(
+    ({ isLoading }) => typeof isLoading === 'boolean' && !isLoading
+  );
+
+  // Initialize current datasource and all active datasources
+  useEffect(() => {
+    if (!allLoaded) {
+      Object.entries(props.datasourceMap).forEach(([datasourceId, datasource]) => {
```

Comment:
In my local test, this seems to go through every index pattern and initialize it. This won't work when running against production which will have potentially many index patterns.

Reply:
This isn't ideal but I don't think it's as bad as what you're suggesting- the index pattern datasource is powered by a single fetch for index patterns.

Reply:
I think we need to flag this and be doubly sure. It seems like a bad idea, and something that won't be caught until run in production or a sizable system, at which point, it's kinda late. Let's make sure to at least bring this up with Q/A.

Reply:
I'd misunderstood this, as Wylie explained in-person. It's doing a paginated load *per* datasource, not per index pattern as I'd been assuming.

Reply:
In case of the indexpattern datasource it loads the first 1000 index patterns with all field information and everything - if there are 1000 metric beat index patterns this could definitely take a while. Maybe we don't have to optimize for this case yet, but I agree it makes sense to bring it up with QA.